openhtf/io/output/json_factory.py

```diff
@@ -44,19 +44,12 @@ def default(self, obj):
   def __call__(self, test_record):
     assert self.filename_pattern, 'filename_pattern required'
     if self.inline_attachments:
-      as_dict = data.ConvertToBaseTypes(test_record)
-      for phase in as_dict['phases']:
+      self.test_record_dict = data.ConvertToBaseTypes(test_record)
+      for phase in self.test_record_dict['phases']:
         for value in phase['attachments'].itervalues():
           value['data'] = base64.standard_b64encode(value['data'])
     else:
-      as_dict = data.ConvertToBaseTypes(test_record, ignore_keys='attachments')
-    if isinstance(self.filename_pattern, basestring):
-      if '{' in self.filename_pattern: 
-        filename = self.filename_pattern.format(**as_dict)
-      else:
-        filename = self.filename_pattern % as_dict
-      with open(filename, 'w') as f:
-        f.write(self.encode(as_dict))
-    else:
-      self.filename_pattern.write(self.encode(as_dict))
-  # pylint: enable=invalid-name
+      self.test_record_dict = data.ConvertToBaseTypes(test_record, ignore_keys='attachments')
+
+    self.data = self.encode(self.test_record_dict)
```

Comment:
stashing values on `self` and then invoking a method that accesses it via `self` is a common mistake (in many languages).  Generally, you want to limit data to as small a namespace as possible (`self` is pretty broad, compared to just the locals of a method).  Passing things in as arguments also keeps a "more functional" interface, where inputs and outputs are more explicit.

the better way to do what you're doing here would be to pass the data into `.Save()`.  Really, though, the even better way is to have the base class' `__call__` method invoke some other method to format the data, and have subclasses override that method.  This reduces code duplication even more (subclasses don't have to all call `Save()` explicitly) and has subclasses override exactly the behavior they want, and nothing more (ideally, you don't want subclasses to have to reimplement whole chunks of logic).